dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+
+        PushNotificationChannelManager channelManager{};
+        winrt::PushNotificationChannel pushChannelReceived{ nullptr };
+
+        for (auto backOffTimeInSeconds = c_minBackoffSeconds; backOffTimeInSeconds <= c_maxBackoffSeconds * 2; backOffTimeInSeconds *= 2)
+        {
+            try
+            {
+                pushChannelReceived = co_await channelManager.CreatePushNotificationChannelForApplicationAsync();
+                channelRequestResult = S_OK;
+                status = PushNotificationChannelStatus::CompletedSuccess;
+                break;
+            }
+            catch (...)
+            {
+
+                auto channelRequestException = hresult_error(to_hresult(), take_ownership_from_abi);
+
+                if ((backOffTimeInSeconds <= c_maxBackoffSeconds) && isChannelRequestRetryable(channelRequestException.code()))
+                {
+                    channelStatus.extendedError = channelRequestException.code();
+                    channelStatus.status = PushNotificationChannelStatus::InProgressRetry;
+                    channelStatus.retryCount = ++retryCount;
+
+                    progress(channelStatus);
+                }
+                else
+                {
+                    channelRequestResult = channelRequestException.code();
+                    status = PushNotificationChannelStatus::CompletedFailure;
+                    break;
+                }
+            }
+
+            co_await winrt::resume_after(std::chrono::seconds(backOffTimeInSeconds));
+        }
+
+
+        if (status == PushNotificationChannelStatus::CompletedSuccess)
+        {
+            // Returns a com_ptr to the implementation type
+            auto pushChannel =
+                winrt::make_self<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationChannel>(pushChannelReceived);
+
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                *(pushChannel.get()),
+                channelRequestResult,
+                status);
+        }
+        else if (status == PushNotificationChannelStatus::CompletedFailure)
+        {
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                nullptr,
+                channelRequestResult,
+                status);
+        }
+
+        co_return channelResult;
+    }
+
+    PushNotificationRegistrationToken PushNotificationManager::RegisterActivator(PushNotificationActivationInfo const& details)
+    {
+        winrt::guid taskClsid = details.TaskClsid();
+        DWORD cookie = 0;
+        BackgroundTaskRegistration registeredTask = nullptr;
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::PushTrigger))
+        {
+            bool taskRegistered = false;
+            for (auto task : BackgroundTaskRegistration::AllTasks())
```

Comment:
API design note about background tasks: The design of background tasks is annoying and forces this ugly pattern on all clients.

But we can simplify the pattern at least.

```cpp
auto tasks = BackgroundRegistration::AllTasks();
taskRegistered = std::any_of(begin(tasks), end(tasks), [&](auto&& task) { return task.Value().Name() == backgroundTaskName; });
```
```